### PR TITLE
Natural light configuration

### PR DESCRIPTION
**Pull request by dengste**
New widget to configure natural light on Kobo Aura One and H2O2.

### CODE REVIEW

**Comment by dengste** [comment]
Further improvements for issue #2285. While fiddling with these values is not for the faint of heart, I'm hoping that people will post some values they find useful and maybe we can digest some nice presets from those.

Screenshot:

nl_widget

**Comment by dengste** [comment]
Not sure what to do about that failing check. IMHO the Size.span.vertical_large was still a bit too small, but I can remove the multiplication with 2, of course.

**Comment by Frenzie** [comment]
I didn't make the check intelligent enough to deal with this scenario. It's just a stupid width/height/a few other things = [0-9]. It'd probably stop complaining if you stuck the *2 on the other side but I haven't checked.

**Comment by Frenzie** [comment]
But there's no objection to adding a larger size either. That whole Size thing is basically my attempt to standardize the semi-randomness we had going on previously. The naming convention I used in font would result in something like Size.span.vertical_x_large.

Looks very nice btw!

File: frontend/ui/widget/frontlightwidget.lua [hunk 1]
```diff
@@ -19,6 +19,7 @@ local UIManager = require("ui/uimanager")
 local VerticalGroup = require("ui/widget/verticalgroup")
 local VerticalSpan = require("ui/widget/verticalspan")
 local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local NaturalLight = require("ui/widget/naturallightwidget")
```
Comment (Frenzie) [inline frontend/ui/widget/frontlightwidget.lua]: Please stick it in alphabetically. See https://github.com/koreader/koreader-base/wiki/Coding-style#order-of-requires :-)

File: frontend/ui/widget/naturallightwidget.lua [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+local LineWidget = require("ui/widget/linewidget")
+local OverlapGroup = require("ui/widget/overlapgroup")
+local Size = require("ui/size")
+local TextBoxWidget = require("ui/widget/textboxwidget")
+local TextWidget = require("ui/widget/textwidget")
+local UIManager = require("ui/uimanager")
+local VerticalGroup = require("ui/widget/verticalgroup")
+local VerticalSpan = require("ui/widget/verticalspan")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local InputText = require("ui/widget/inputtext")
```
Comment (Frenzie) [inline frontend/ui/widget/naturallightwidget.lua]: Same remark as above.

File: frontend/ui/widget/naturallightwidget.lua [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+    self.powerd.fl.exponent = self.exponent[2]:getText()
+    self.powerd.fl:setNaturalBrightness()
+end
+
+-- Create an InputText with '-' and '+' button next to it. Tapping
+-- those buttons will de/increase by 'step', and 'step/10' on hold.
+function NaturalLightWidget:adaptableNumber(initial, step)
+    local minus_number_plus = HorizontalGroup:new{ align = "center" }
+    local input_text = InputText:new{
+        parent=self,
```
Comment (Frenzie) [inline frontend/ui/widget/naturallightwidget.lua]: Total nitpick, but could you write it with spaces? `parent = self` etc.

File: frontend/ui/widget/naturallightwidget.lua [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+                self.powerd.fl.exponent}
+end
+
+function NaturalLightWidget:createFrame()
+    self.nl_title = FrameContainer:new{
+        padding = Size.padding.default,
+        margin = Size.margin.title,
+        bordersize = 0,
+        TextWidget:new{
+            text = _("Natural Light configuration"),
```
Comment (Frenzie) [inline frontend/ui/widget/naturallightwidget.lua]: Probably just `Natural light configuration`. (Or `Natural Light Configuration`.)

File: frontend/ui/widget/naturallightwidget.lua [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+    local vspan = VerticalSpan:new{ width = 2 * Size.span.vertical_large }
+    local vertical_group = VerticalGroup:new{ align = "center" }
+    local title_group = HorizontalGroup:new{ align = "center" }
+    local white_group = HorizontalGroup:new{ align = "center" }
+    local red_group = HorizontalGroup:new{ align = "center" }
+    local green_group = HorizontalGroup:new{ align = "center" }
+    local exponent_group = HorizontalGroup:new{ align = "center" }
+    local button_group = HorizontalGroup:new{ align = "center" }
+    local text_gain = TextBoxWidget:new{
+        text = _("Gain"),
```
Comment (Frenzie) [inline frontend/ui/widget/naturallightwidget.lua]: Would it be clearer as amplification?

File: frontend/ui/widget/naturallightwidget.lua [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+        radius = 0,
+        width = self.width * 0.2,
+        show_parent = self,
+        callback = function()
+            self:setAllValues(self.old_values)
+            self:onClose()
+        end,
+    }
+    local button_ok = Button:new{
+        text = "OK",
```
Comment (Frenzie) [inline frontend/ui/widget/naturallightwidget.lua]: Is there anything better we can say here? E.g., `Save` or `Save settings`. See the UX guidelines here: http://koreader.rocks/doc/modules/ui.widget.confirmbox.html (not a requirement but nice to have)

**Comment by dengste** [comment]
Thanks for the review. I've added two fixups, please squash them if they're OK.

**Comment by Frenzie** [comment]
Thanks! How come you don't want to squash 'em into the relevant commits yourself? :-)

**Comment by dengste** [comment]
If I squash them and force-push, I thought it would be harder for you to see what I've actually changed? But of course I can do that in the future if you prefer that.

**Comment by Frenzie** [comment]
No, you're quite right about that. I meant after it's all done. I do prefer a "perfect" PR that I can just rebase or merge[1] but mostly I'm just curious about the rationale.

[1] That's half a minute I could've spent singing your praises in the release notes for the next GH release instead. /grins

**Pull request merged**